File: src/controllers/public.js
const path = require('path');
const { listPosts, getPostBySlug } = require('../utils/postStore');
const { buildRssFeed } = require('../utils/rss');
const { siteUrl, blogTitle } = require('../config');
const { PUBLIC_DIR } = require('../utils/paths');
const { recordView } = require('../utils/analytics');

async function showBlogs(req, res, next) {
    try {
        const posts = await listPosts();
        res.render('index', { title: blogTitle, posts });
    } catch (error) {
        next(error);
    }
}

function showPortfolio(req, res) {
    res.render('portfolio', { title: 'Portfolio · Srinath' });
}

async function showBlogPost(req, res, next) {
    try {
        const post = await getPostBySlug(req.params.slug, { includeDrafts: false });
        if (!post || post.status !== 'published') {
            return res.status(404).render('404', { title: 'Not found' });
        }
        recordView(post.slug, req);
        const allPosts = await listPosts();
        const idx = allPosts.findIndex(p => p.slug === post.slug);
        const prevPost = idx < allPosts.length - 1 ? allPosts[idx + 1] : null;
        const nextPost = idx > 0 ? allPosts[idx - 1] : null;
        res.render('post', { title: post.title, post, prevPost, nextPost });
    } catch (error) {
        next(error);
    }
}

function showLanding(req, res) {
    res.render('landing', { title: res.locals.blogTitle });
}

async function showDownloadResume(req, res, next) {
    try {
        const resumePath = path.join(PUBLIC_DIR, 'resume', 'srinathShresthaResume.pdf');
        return res.download(resumePath, 'SrinathShresthaResume.pdf');
    } catch (error) {
        console.error(error);
        return res.status(500).send('Error downloading resume');
    }
}

async function showRssFeed(req, res, next) {
    try {
        const posts = await listPosts();
        const detailed = await Promise.all(
            posts.filter((post) => post.status === 'published')
                .map((post) => getPostBySlug(post.slug))
        );
        const xml = buildRssFeed({
            siteUrl,
            blogTitle,
            posts: detailed.filter(Boolean),
        });
        res.type('application/rss+xml').send(xml);
    } catch (error) {
        next(error);
    }
}

module.exports = {
    showLanding,
    showBlogs,
    showPortfolio,
    showDownloadResume,
    showBlogPost,
    showRssFeed,
};

File: routes/public.py
from pathlib import Path
from fastapi import APIRouter, Request
from fastapi.responses import FileResponse, Response
import post_store as store
from rss import build_rss
from config import cfg
from templates_env import tmpl

router = APIRouter()
_RESUME = Path(__file__).parent.parent / "public" / "resume" / "srinathShresthaResume.pdf"


@router.get("/")
async def landing(req: Request):
    return tmpl.TemplateResponse(req, "landing.html", {"title": cfg.BLOG_TITLE})


@router.get("/blogs")
async def blogs(req: Request):
    posts = await store.list_posts()
    return tmpl.TemplateResponse(req, "index.html", {"title": cfg.BLOG_TITLE, "posts": posts})


@router.get("/blogs/{slug}")
async def blog_post(slug: str, req: Request):
    post = await store.get_post(slug)
    if not post or post["status"] != "published":
        return tmpl.TemplateResponse(req, "404.html", {"title": "Not found"}, status_code=404)
    return tmpl.TemplateResponse(req, "post.html", {"title": post["title"], "post": post})


@router.get("/portfolio")
async def portfolio(req: Request):
    return tmpl.TemplateResponse(req, "portfolio.html", {"title": f"Portfolio · {cfg.BLOG_TITLE}"})


@router.get("/resume")
def resume():
    return FileResponse(_RESUME, filename="SrinathShresthaResume.pdf")


@router.get("/rss.xml")
async def rss():
    posts = await store.list_posts()
    detailed = [await store.get_post(p["slug"]) for p in posts if p["status"] == "published"]
    xml = build_rss(cfg.SITE_URL, cfg.BLOG_TITLE, [p for p in detailed if p])
    return Response(xml, media_type="application/rss+xml")


@router.get("/health")
def health():
    return {"ok": True}
